std/meta/index.zig

```diff
@@ -0,0 +1,691 @@
+    return @TagType(T)(u);
+}
+
+test "std.meta.activeTag"
+{
+    const UE = enum.
+    {
+        Int,
+        Float,
+    };
+    
+    const U = union(UE).
+    {
+        Int: u32,
+        Float: f32,
+    };
+    
+    var u = U.{ .Int = 32, };
+    debug.assert(activeTag(u) == UE.Int);
+    
+    u = U.{ .Float = 112.9876, };
+    debug.assert(activeTag(u) == UE.Float);
+
+}
+
+//see: bytesOf
+fn BytesReturnType(comptime P: type) type
+{
+    return if(trait.isConstPtr(P)) []const u8 else []u8;
+}
+///Given a pointer to a single item, returns a slice of the underlying bytes, preserving the
+/// pointer's const-ness.
+pub fn bytes(ptr: var) BytesReturnType(@typeOf(ptr))
+{
+    const P = @typeOf(ptr);
+    if(comptime !trait.isSingleItemPtr(P)) @compileError("meta.bytes requires a single item " 
+        ++ "pointer,  passed " ++ @typeName(P));
+    const T = Child(P);
+    
+    const ReturnType = comptime BytesReturnType(P);
+    const IntermediateType = comptime (if(trait.isConstPtr(P)) [*]const u8 else [*]u8);
+    
+    var slice: ReturnType = undefined;
+    slice.ptr = @ptrCast(IntermediateType, ptr);
+    slice.len = @sizeOf(T);
+    return @sliceToBytes(slice);
+}
+
+test "std.meta.bytes"
+{
+    const deadbeef = u32(0xDEADBEEF);
+    const deadbeef_bytes = switch(builtin.endian)
+    {
+        builtin.Endian.Big => "\xDE\xAD\xBE\xEF",
+        builtin.Endian.Little => "\xEF\xBE\xAD\xDE",
+    };
+    
+    debug.assert(std.mem.eql(u8, bytes(&deadbeef), deadbeef_bytes));
+    
+    var codeface = u32(0xC0DEFACE);
+    for(bytes(&codeface)) |*b| b.* = 0;
+    debug.assert(codeface == 0);
+    
+    const S = packed struct.
+    {
+        a: u8,
+        b: u8,
+        c: u8,
+        d: u8,
+    };
+    
+    const inst = S.{ .a = 0xBE, .b = 0xEF, .c = 0xDE, .d = 0xA1, };
+    debug.assert(std.mem.eql(u8, bytes(&inst), "\xBE\xEF\xDE\xA1"));
+}
+
+/// Provides the size of a type padded to the nearest byte, instead of however the
+///  compiler actually pads it. Useful for determining the size of a type as it would
+///  likely be stored to disk.
+pub fn byteAlignedSizeOf(comptime T: type) usize
+{
+    comptime
+    {
+        switch(@typeId(T))
+        {   
+            builtin.TypeId.Bool,
+            builtin.TypeId.Int,
+            builtin.TypeId.Float => return @sizeOf(T),
+            builtin.TypeId.Enum => return @sizeOf(@TagType(T)),
+            builtin.TypeId.Array =>
+            {
+                return byteAlignedSizeOf(Child(T)) * T.len;
+            },
+            builtin.TypeId.Struct =>
+            {
+                const info = @typeInfo(T).Struct;
+                
+                if(info.layout == builtin.TypeInfo.ContainerLayout.Packed) return @sizeOf(T);
+                
+                var size = usize(0);
+                for(info.fields) |field|
+                {
+                    size += byteAlignedSizeOf(field.field_type);
+                }
+                return size;
```

Comment:
ContainerLayout.Auto will not have this size. The fields will be reordered (they don't currently) depending on how the compiler decides to optimize the struct.

Reply:
I don't understand the objection. The purpose of the function is specifically to ignore how the compiler lays out structs in memory so that they can be written and read to disk/sockets/whatever in a consistent way.

Reply:
I'm not sure I understand why this function helps with this usecase. Wouldn't you just do this loop?
```
inline for (fileds) |field| {
    stream.write(toBytes(field));
    stream.writeNTimes(0x0, @sizeOf(@typeOf(field)) - alignment(@typeOf(field));
}

```

Why do you need a "fake" size for this usecase?

Reply:
And if the field is itself a struct?

On the one hand, the compiler reserves the right to layout and pad a struct however it wants unless it is packed, which is good for optimization. On the other hand, when working with file formats (for instance), structs *do* have a specific layout, and endianess, and often do not waste space on padding. This function, when combined with another, allows me to let the compiler choose an optimal layout in memory, yet still read and write that structure to disk easily.

Perhaps I should remove it for now and include it in a later commit when I add the generic type read/write functions that would make its usefulness clear.

Reply:
I think that is a good idea. That'll make its use case clear